### ISSUE TITLE
KaTex Math render problem

### ISSUE THREAD

**Issue by upupming** [neilsustc/vscode-markdown#303]
#### What is the problem?

This is a small problem. This code:

```
$\text{\`{a}}$ <code>\\`{a}</code>
```
on [katex.org](https://katex.org/docs/supported.html) should be rendered like this:

![image](https://user-images.githubusercontent.com/24741764/47085900-a4da8f80-d24a-11e8-8445-5619f6b3ff5b.png)

However, I got this:

![image](https://user-images.githubusercontent.com/24741764/47086140-4bbf2b80-d24b-11e8-8468-ce906ae4f9b6.png)

#### How can I reproduce it?

Write in your md file:

```
|||||
|:---------------------|:---------------------|:---------------------|:-----
|$\text{\'{a}}$ `\'{a}`|$\text{\~{a}}$ `\~{a}`|$\text{\.{a}}$ `\.{a}`|$\text{\H{a}}$ `\H{a}`
|$\text{\`{a}}$ <code>\\`{a}</code>|$\text{\={a}}$ `\={a}`|$\text{\"{a}}$ `\"{a}`|$\text{\v{a}}$ `\v{a}`
|$\text{\^{a}}$ `\^{a}`|$\text{\u{a}}$ `\u{a}`|$\text{\r{a}}$ `\r{a}`|

```

#### Is there any error message in the console?

No.

**Comment by upupming** [comment]
Oh, it seems we should add another `\`(to https://raw.githubusercontent.com/Khan/KaTeX/master/docs/supported.md):

```
<code>\\\`{a}</code>
```

**Comment by neilsustc** [comment]
It is all about escaping the special characters `\` and <code>`</code>.

For **inline code span**, Markdown parser does seem to escape them. So you need to use `\\` to denote `\`.
And for **code block**, it seems that we can use `\` without escaping.

**Comment by upupming** [comment]
I find math rendering in `vscode-markdown` is really powerful! Thanks for your great work, and based on your project, I create [a plugin](https://github.com/upupming/docsify-katex) for rendering KaTex in docsify, hope it helps.

**Comment by upupming** [comment]
Hi, I found this problem can be solved by setting `html` option to `true` when using `markdown-it`:

```js
markdownit({html: true})
  .use(markdownitKatex, { "throwOnError": false });
```

In this way, the `\` in `code` tag will not be escaped and <code>`</code> will be escaped.

For example:

```
<code>`</code>
```

will be rendered as:

<code>`</code>

However, the following:

```
<code>\</code>
```

will be rendered as:

![image](https://user-images.githubusercontent.com/24741764/47289913-7cc6a400-d62f-11e8-8425-1ba863e577c4.png)

**Comment by neilsustc** [comment]
The `html` option is already `true`. (See the [source code](https://github.com/Microsoft/vscode/blob/master/extensions/markdown-language-features/src/markdownEngine.ts#L40) (for vscode preview), [source code](https://github.com/neilsustc/vscode-markdown/blob/master/src/print.ts#L23) (for exported HTML))

---

```markdown
<code>`</code>
```

looks no problem because you are lucky that there is no following <code>\`</code> character to form an inlne code span. If you try

```markdown
<code>`</code> and `hello`
```

Observe the above Markdown syntax highlight and the markdown-it demo

![image](https://user-images.githubusercontent.com/7588612/47292751-e29b3780-d5ff-11e8-9a27-f92c3bbd9985.png)
![image](https://user-images.githubusercontent.com/7588612/47292807-10807c00-d600-11e8-9c8b-aa0a89776442.png)

both ` and \ need to escape

```markdown
<code>\`</code> and `hello`
```

So theoretically, it should be <code>&lt;code>\\\\\\\`&lt;/code></code> rather than <code>&lt;code>\\\\\`&lt;/code></code>

**Comment by upupming** [comment]
Thanks a lot for your plain explanation!

**Comment by upupming** [comment]
I find another problem with `\neq`, with following equation:

```latex
$$

a \neq b

$$
```

is rendered as:

![20181029165836.png](https://i.loli.net/2018/10/29/5bd6cbc0600c1.png)

it should be rendered as:

![image](https://user-images.githubusercontent.com/24741764/47639347-003a4500-db9c-11e8-917d-7e067a532998.png)

**Comment by neilsustc** [comment]
I also noticed it and suspects it to be a vscode issue. Let's tract it in #252.

**Comment by RoyiAvital** [comment]
I encountered another problem with KaTeX rendering (Though I guess it is a parser thing).

This will work:

```markdown
Does this ${c}^{2} = {a}^{2} + {b}^{2}$ work?
```

While this won't:

```markdown
Does this $ {c}^{2} = {a}^{2} + {b}^{2} $ work?
```

Mind the Gap (Space) after and before the `$` mark.
Could you please let the parser identify ti as Math?

I know many parses do that in order to prevent cases of using the USD sign.
Yet in LaTeX it is acceptable hence to keep compatibility I'd let the option do it here as well.

Even if it is an **Opt In** option in preferences.
So people who asks for LaTeX compatibility will know to use this option and escape the USD sign.

Thank You.

**Comment by yzhang-gh** [comment]
> I know many parses do that in order to prevent cases of using the USD sign.

You are right. https://github.com/waylonflinn/markdown-it-katex#syntax

I am afraid I cannot do much here.

**Comment by RoyiAvital** [comment]
@yzhang-gh , Is that the Math Parser you're using for Math?

**Comment by yzhang-gh** [comment]
Yes

**Comment by RoyiAvital** [comment]
Maybe you can use https://github.com/classeur/markdown-it-mathjax for MathJaX.
Also in order to support spaces after `$` all needed is to edit this line: https://github.com/classeur/markdown-it-mathjax/blob/master/markdown-it-mathjax.js#L77

**Comment by yzhang-gh** [comment]
Of course. Almost all issues can be solved programmatically.

There are various features people would love to add, like many [Markdown dialects](https://github.com/yzhang-gh/vscode-markdown/issues?q=is%3Aissue+is%3Aopen+label%3A%22markdown+dialect%22), different TOC link styles (GitHub, GitLab etc.). You request (MathJax, `$` rule) is one of them.

As a maintainer, I can only do some of the high-priority features (with many votes), and avoid modifying upstream packages which will increase the maintainance effort (some of them have not been updated since 2016).

---

If you really want to solve a problem, there are always 3 ways.
- do it yourself (although a single person's ability is usually limited)
- find many people who also have this problem, hopefully one of them can get it resolved
- find a workaround (change the problem itself)

I am not saying which one is better than the others. But I am sure not all issues can be resolved. (Look VSCode's issue page)
Welcome to the real world. Good luck.

**Comment by RoyiAvital** [comment]
My friend,
I didn't come with demands.
No need for a wake up call. I know reality just as you.

I suggested a feature, I gave a way to do both (MathJaX support and supporting space after `$`).
It is up to you what to do with it.

My logic is addressing Students and other people using LaTeX.
If you feel this ignite motivation on your side, perfect.
If not, I thank you for what you did so far.

As for me, I suggested a contribution of ~Hour of coding.
Not because it is an hour work but in order to show I do appreciate your time and would like to show it.

Again, thank you for this nice extension.
On my side, given a full LaTeX support it would have been the perfect student replacement for LaTex.
